internal/throttling/throttling.go

```diff
@@ -0,0 +1,249 @@
+* replace old limiters
+* guard against concurrency?
+*/
+
+var (
+	//go:embed lua/gcra.lua
+	gcraLua         string
+	gcraRedisScript *redis.Script
+	//go:embed lua/sortedset.lua
+	sortedSetLua    string
+	sortedSetScript *redis.Script
+)
+
+func init() {
+	gcraRedisScript = redis.NewScript(gcraLua)
+	sortedSetScript = redis.NewScript(sortedSetLua)
+}
+
+type redisSpeaker interface {
+	redis.Scripter
+	redisSortedSetRemover
+}
+
+type TokenReturner interface {
+	Return(context.Context) error
+}
+
+type statsCollector interface {
+	NewTaggedStat(name, statType string, tags stats.Tags) stats.Measurement
+}
+
+type Limiter struct {
+	// for Redis configurations
+	// a default redisSpeaker should always be provided for Redis configurations
+	redisSpeaker redisSpeaker
+	// redisToClientMap does not require a Mutex because it is only read from and written once in the constructor
+	redisKeyToClientMap map[string]redisSpeaker
+
+	// for in-memory configurations
+	gcra   *gcra
+	goRate *goRate
+
+	// other flags
+	useGCRA            bool
+	useGCRABurstAsRate bool
+	useGoRate          bool
+
+	// metrics
+	statsCollector statsCollector
+}
+
+func New(options ...Option) (*Limiter, error) {
+	rl := &Limiter{}
+	for i := range options {
+		options[i].apply(rl)
+	}
+
+	if rl.statsCollector == nil {
+		rl.statsCollector = stats.Default
+	}
+
+	if rl.redisSpeaker != nil {
+		if rl.useGoRate {
+			return nil, fmt.Errorf("redis and go-rate are mutually exclusive")
+		}
+		return rl, nil
+	}
+	if len(rl.redisKeyToClientMap) > 0 {
+		return nil, fmt.Errorf("redis key to client map is configured without a default client")
+	}
+
+	switch {
+	case rl.useGCRA:
+		rl.gcra = &gcra{}
+	default:
+		rl.goRate = &goRate{}
+	}
+	return rl, nil
+}
+
+// Limit returns true if the limit is not exceeded, false otherwise.
+func (l *Limiter) Limit(ctx context.Context, cost, rate, window int64, key string) (
+	bool, TokenReturner, error,
+) {
+	if cost < 1 {
+		return false, nil, fmt.Errorf("cost must be greater than 0")
+	}
+	if rate < 1 {
+		return false, nil, fmt.Errorf("rate must be greater than 0")
+	}
+	if window < 1 {
+		return false, nil, fmt.Errorf("window must be greater than 0")
+	}
+	if key == "" {
+		return false, nil, fmt.Errorf("key must not be empty")
+	}
+	switch {
+	case l.useGCRA:
+		if l.redisSpeaker != nil {
+			defer l.getTimer(key, "redis-gcra", rate, window)()
+			return l.redisGCRA(ctx, cost, rate, window, key)
+		}
+		defer l.getTimer(key, "gcra", rate, window)()
+		return l.gcraLimit(ctx, cost, rate, window, key)
+	case l.redisSpeaker != nil:
+		defer l.getTimer(key, "redis-sorted-set", rate, window)()
+		return l.redisSortedSet(ctx, cost, rate, window, key)
+	default:
+		defer l.getTimer(key, "go-rate", rate, window)()
+		return l.goRateLimit(ctx, cost, rate, window, key)
+	}
+}
+
+func (l *Limiter) redisSortedSet(ctx context.Context, cost, rate, window int64, key string) (
+	bool, TokenReturner, error,
+) {
+	redisSpeaker := l.getRedisSpeaker(key)
```

Comment:
Our current usage patterns don't justify supporting a different redis client per key. For the time being I believe we could live without a `redisKeyToClientMap` and revisit once it is actually needed.

Reply:
As a side note, if this is to support global redis limiters in the future, then we still don't need support for multiple redis clients within the same `Limiter`, since each router will either need a local or global limiter, not both at the same time

Reply:
We still haven't decided whether we want to pay for a big AWS Elasticache to take care of *all* requests across *all* clusters (mt, pro, enterprise) or not. We have a call on Wednesday to discuss that. My take is that 1) it would cost too much 2) SPOF.
I'd go with a generic AWS Elasticache for the "global" destinations that share limits across all clusters and then have cheap standalone Redis masters deployed in the same K8s cluster.
Let's see what comes out of the call and we'll decide whether to keep this map afterwards.